### PR TITLE
ESLint v8 (#35576)

### LINKED ISSUE #35576: [@wordpress/eslint-plugin] Support for ESlint v8
### Description

ESlint v8 [has been released](https://eslint.org/blog/2021/10/eslint-v8.0.0-released) and it includes a few breaking changes.

`@wordpress/eslint-plugin` should support it. At the very least it should:

- Declare `8` as a valid peer dependency
- Update 3rd party `eslint-plugin-*` dependencies to a version that supports ESLint v8

I haven't check if the provided rules must be updated as well to work in ESLint v8

### Step-by-step reproduction instructions

N/A

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by simonhammes**
## Description

Closes https://github.com/WordPress/gutenberg/issues/35576.

#### @wordpress/eslint-plugin
- Updated packages for compatibility with ESLint v8

#### @wordpress/scripts
- Updated ESLint to v8 (#35576)
  Changelog: https://eslint.org/docs/user-guide/migrating-to-8.0.0

## How has this been tested?
- `npm run lint`  
   The parsing errors will most likely be gone once #36244 is merged (the optional Babel plugin was enabled by default since 7.14, see https://github.com/babel/babel/pull/13232)

I only just noticed that @gziolo was working on this, so some changes in this PR overlap with changes in #36244.

I will also add changelog entries before this gets merged.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/api-fetch/src/test/index.js [hunk 1]
```diff
@@ -127,6 +127,7 @@ describe( 'apiFetch', () => {
 		);
 
 		return apiFetch( { path: '/random' } ).catch( ( body ) => {
+			// eslint-disable-next-line jest/no-conditional-expect
```
Comment (simonhammes) [inline packages/api-fetch/src/test/index.js:130]: The `jest/no-condtional-expect` rule now also covers catching promises, so I had to add these comments to disable the rule.

Comment (simonhammes) [reply packages/api-fetch/src/test/index.js:130]: Link: https://github.com/jest-community/eslint-plugin-jest/pull/819

File: packages/block-serialization-default-parser/test/index.js [hunk 2]
```diff
@@ -17,7 +17,7 @@ import {
  */
 import { parse } from '../src';
 
-describe( 'block-serialization-default-parser-js', jsTester( parse ) ); // eslint-disable-line jest/valid-describe
+describe( 'block-serialization-default-parser-js', jsTester( parse ) ); // eslint-disable-line jest/valid-describe-callback
```
Comment (simonhammes) [inline packages/block-serialization-default-parser/test/index.js:20]: The rule name changed: https://github.com/jest-community/eslint-plugin-jest/releases/tag/v25.0.0

File: packages/env/lib/config/test/config.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-/* eslint-disable jest/no-try-expect, jest/no-conditional-expect */
+/* eslint-disable jest/no-conditional-expect */
```
Comment (simonhammes) [inline packages/env/lib/config/test/config.js:1]: The `jest/no-try-expect` rule was removed with v25: https://github.com/jest-community/eslint-plugin-jest/releases/tag/v25.0.0

File: packages/eslint-plugin/rules/data-no-store-string-literals.js [hunk 4]
```diff
@@ -198,6 +198,7 @@ function getFixes( fixer, context, callNode ) {
 module.exports = {
 	meta: {
 		type: 'problem',
+		hasSuggestions: true,
```
Comment (simonhammes) [inline packages/eslint-plugin/rules/data-no-store-string-literals.js:201]: This is now necessary (since ESLint v8) if a rule provides suggestions: https://eslint.org/docs/user-guide/migrating-to-8.0.0

**Comment by gziolo** [comment]
Thank you for starting this PR. 

> I only just noticed that @gziolo was working on this, so some changes in this PR overlap with changes in #36244.

I expect this PR will land this week. I will let you know when it happens so you could rebase this one.

File: packages/eslint-plugin/configs/recommended.js [hunk 5]
```diff
@@ -21,7 +21,6 @@ const config = {
 	extends: [
 		require.resolve( './recommended-with-formatting.js' ),
 		'plugin:prettier/recommended',
-		'prettier/react',
```
Comment (gziolo) [inline packages/eslint-plugin/configs/recommended.js:24]: Can you explain why this line should be removed?

**Review by gziolo — COMMENTED** [review]
This is looking good. We still need to document all the major version updates for packages as breaking changes. We have some fixes included in this PR which confirm that upgrade for the new version of the `@wordpress/eslint-plugin` won't be seamless.

Let's see which errors will still require some action after merging changes for Babel v7.16.0.

File: packages/eslint-plugin/configs/recommended.js (hunk 5 above)
Comment (simonhammes) [reply packages/eslint-plugin/configs/recommended.js:24]: This is the relevant changelog entry: https://github.com/prettier/eslint-config-prettier/blob/5a2f0e24382a8e1322784bc836d47706cccd2f1e/CHANGELOG.md#version-800-2021-02-21

I also documented this in the [commit message](https://github.com/WordPress/gutenberg/pull/36283/commits/60aa8bb40e74f4e91f53746a4d9e1cce2ede293d).

**Comment by simonhammes** [comment]
> This is looking good. We still need to document all the major version updates for packages as breaking changes. 

I will document those changes :+1:

**Comment by gziolo** [comment]
I landed https://github.com/WordPress/gutenberg/pull/36244. Let’s see if rebasing is going to be enough to make all Continues Integration checks pass.

**Comment by simonhammes** [comment]
> I landed #36244. Let’s see if rebasing is going to be enough to make all Continues Integration checks pass.

I rebased this PR.  
Unfortunately, all eslint-plugin tests are still failing: https://github.com/WordPress/gutenberg/runs/4180615328?check_suite_focus=true#step:5:908

I tried to debug the issue and they might be related to the following issues/PRs:
- Same error message: https://github.com/eslint/eslint/issues/14936#issuecomment-905786726
- https://github.com/eslint/eslintrc/pull/57 => https://github.com/eslint/eslint/pull/15164

It seems like the testing failures arise from an interaction between Jest & the upgrade from ESLint from v7 to v8. 

I suspect that the following PR made to ESLint is the main culprit: https://github.com/eslint/eslint/pull/14706 (since Jest does not seem to support the `exports` field in `package.json`: https://github.com/facebook/jest/issues/9771).

I could get the tests for the eslint-plugin to pass locally by adding an `/** @jest-environment node */` annotation (to force Jest to use the `node` environment instead of the default `jsdom` environment) to the top of each test file in the `packages/eslint-plugin/rules/__tests__` directory **and** running the tests directly with Jest (command: `node node_modules/.bin/jest packages/eslint-plugin/`). But this is not a _real solution_.

I also tried to upgrade ESLint to v8 on this branch: https://github.com/WordPress/gutenberg/pull/33287 to eliminate the possibility of a possible incompatibility between ESLint v8 and Jest v26, but the tests also failed with the same error message as in the linked CI run at the top of this comment.

**Comment by gziolo** [comment]
It looks like there are some incompatible changes introduced between ESLint v8 and Jest v27 that will force us to update both tools at a similar time. I also encountered several issues when trying to upgrade Jest to v27, but I managed to resolve most of them. The only blocker is related to React Native so we need to wait until the Mobile team brings React Native to the more recent version. Do you think we will have to wait until Jest v27 is there before we can upgrade ESLint to v8?

**Comment by simonhammes** [comment]
@gziolo I managed to solve the issue with the failing unit tests.

However, the "Static Analysis" job is failing because of an uncommitted change to `package-lock.json`.

The job was failing earlier, then I ran npm install locally and committed the change: https://github.com/WordPress/gutenberg/pull/36283/commits/920598e4ed38b99f86f7b9c120d5a65d4f107900

Now the job is failing with the exact same **opposite diff**.

**Comment by gziolo** [comment]
> I managed to solve the issue with the failing unit tests.

Great news! 🎉 Awesome work 👏

> The job was failing earlier, then I ran npm install locally and committed the change: 920598e

Yes, I’m seeing it a lot recently. We really need to upgrade to npm 8 once the next major WordPress release (5.9 is scheduled for January) is out.

I think you can manually revert the changes in the lock file that get reported in the CI job.

File: package-lock.json [hunk 6]
```diff
@@ -22019,16 +22092,6 @@
 						"yauzl": "^2.7.0"
 					},
 					"dependencies": {
```
Comment (gziolo) [inline package-lock.json]:
```suggestion
					"dependencies": {
						"are-we-there-yet": {
 							"version": "2.0.0",
 							"resolved": "https://registry.npmjs.org/are-we-there-yet/-/are-we-there-yet-2.0.0.tgz",
 							"integrity": "sha512-Ci/qENmwHnsYo9xKIcUJN5LeDKdJ6R1Z1j9V/J5wyq8nh/mYPEpIKJbBZXtZjG04HiK7zV/p6Vs9952MrMeUIw==",
 							"dev": true,
 							"requires": {
 								"delegates": "^1.0.0",
 								"readable-stream": "^3.6.0"
 							}
 						},
```

Comment (gziolo) [reply package-lock.json]: I hope this is all we need to make CI happy.

File: packages/eslint-plugin/package.json [hunk 7]
```diff
@@ -31,25 +31,25 @@
 	],
 	"main": "index.js",
 	"dependencies": {
-		"@babel/eslint-parser": "^7.16.0",
-		"@typescript-eslint/eslint-plugin": "^4.31.0",
-		"@typescript-eslint/parser": "^4.31.0",
+	  	"@babel/eslint-parser": "^7.16.0",
```
Comment (gziolo) [inline packages/eslint-plugin/package.json]: Nit: I noticed some mixed tabs and spaces:
```suggestion
		"@babel/eslint-parser": "^7.16.0",
```

File: test/unit/jest.config.js [hunk 8]
```diff
@@ -14,6 +14,8 @@ module.exports = {
 		[ `@wordpress\\/(${ transpiledPackageNames.join(
 			'|'
 		) })$` ]: 'packages/$1/src',
+		// See https://github.com/facebook/jest/issues/11100#issuecomment-967161978
+		'@eslint/eslintrc': '@eslint/eslintrc/dist/eslintrc-universal.cjs',
```
Comment (gziolo) [inline test/unit/jest.config.js]: Should we include this workaround in the Jest preset that is shared also with `@wordpress/scripts`:

https://github.com/WordPress/gutenberg/blob/trunk/packages/jest-preset-default/jest-preset.js

I'm also not sure when this issue happens, so maybe it's only specific to the Gutenberg setup.

**Comment by gziolo** [comment]
I guess we can land this PR pretty soon once the remaing issues/questions get addressed.

File: test/unit/jest.config.js (hunk 8 above)
Comment (simonhammes) [reply test/unit/jest.config.js]: I've put this workaround in the jest-preset-default package since all consumers of the preset will now benefit from it.
The unit tests still run (on my machine), hopefully CI agrees.

The issue should happen whenever someone tries to import eslint (v8) inside of Jest test files (e.g. when using `ruleTester`).

File: package-lock.json (hunk 6 above)
Comment (simonhammes) [reply package-lock.json]: GitHub doesn't let me apply your suggestion...
The browser console shows an error whenever I click the "Commit suggestion" button.

Comment (simonhammes) [reply package-lock.json]: I applied this patch locally: fdf6975

File: test/unit/jest.config.js (hunk 8 above)
Comment (gziolo) [reply test/unit/jest.config.js]: It’s rather an edge case then, but it’s something that Jest 27 resolves so it’s fine to keep the workaround for now 👍

**Review by gziolo — APPROVED** [review]
Great work with the ESLint upgrade. Everything looks good, and it seems to work correctly. Let’s land it as soon as Continuous Integration turns green.

**Comment by WraithKenny** [comment]
any news

**Comment by gziolo** [comment]
It's now available for testing on npm with:

```bash
npm install --save-dev @wordpress/eslint-plugin@next
```

or

```bash
npm install --save-dev @wordpress/scripts@next
```

The stable version should be available in the second half of January.
